### ISSUE TITLE
About RoCC (Rocket Custom Coprocessor)

### ISSUE THREAD

**Issue by JongyCysec** [kaist-cp/cs492-uarch#88]
# 1. Materials from lecture note
- Gemmini instructions are interspersed among standard CPU instructions
- Gemmini does not fetch its own instructions. When the CPU encounters a RoCC instruction, it:
  1) Checks the opcode to identify the target
  2) Dispatches the instruction to the appropriate co-processor
  3) Immediately resumes CPU execution
  4) Synchronizes with fences as necessary

![image](https://github.com/user-attachments/assets/4e41fb23-7838-4145-97b2-83e2853d0f10)

# 2. Materials from risc-v rocket chip tutorial, 
![image](https://github.com/user-attachments/assets/78e6b724-2f0f-4a7e-90f9-30dc4576bac7)

# Question 1
- Then, if we compile AI model using AI compiler like TVM, XLA then does AI compiler generate assembly code that  includes both standard Rocket core instructions and custom Gemmini instructions?

- In contrast, when targeting a GPU, I expect that AI compilers  generate CUDA kernels for GPU-specific operations with host CPU code to orchestrate data movement and kernel execution.

- In other words, I think that there is no need for a separate programming model like CUDA when targeting rocket chip with gemmini. Everything is expressed within the RISC-V ISA, using the Rocket core as the "host" CPU.

May I request you to check the validity of my expectation above?

# Question 2. 
Suppose we're going to implement Rocc functionality based on 5-stage pipeline riscv cpu that we've gone through while doing homework 2&3.

So I contemplated on what should be additionally added, modified, and considered to implement Rocc fuctionality. 

Of course, additional decoding logic for interpreting extended ISA for Rocc is required in decode stage. 
But I think that when cpu dispatches the extended instruction to co-processor matters.

### Issue. When to dispatches the instruction to the appropriate co-processor
- If decode stage dispatches the instruction to co-processor and becomes idle until receiving response from the co-processor, then I wonder whether we can appropriately handle redirect signal from execute stage or mem stage. That is because, such extended instruction for co-processor is not directed to execute stage but instead it is dispatched to co-processor. 

- It is said in the lecture note that **all previous memory accesses will be completed before executing subsequent instructions**. Therefore, I think such extended instruction for co-processor should be dispatched after memory stage. 

I would be very thankful if I get some related materials regarding this issues or extra considerations that are needed to implement Rocc functionality on traditional pipeline 5-stage risc-v cpu.

**Comment by woojinnn** [comment]
First, I'll answer only what I know. I'll study a little bit more, and I'll add more.
## Question 1.
> Then, if we compile AI model using AI compiler like TVM, XLA then does AI compiler generate assembly code that includes both standard Rocket core instructions and custom Gemmini instructions?

Yes, I think so. Maybe below materials (targeting the integration between TVM and RoCC) would help.
* https://carrv.github.io/2020/papers/CARRV2020_paper_3_Xu.pdf
* https://discuss.tvm.apache.org/t/presenting-the-generation-of-code-for-the-gemmini-accelerator/14107
* https://arxiv.org/abs/2212.03034

## Question 2.
I think it's after memory stage, as you mentioned in the second bullet point.
From the 8th slide of [Generating Rocket/BOOM SoCs with Rocket Chip (Hot Chips 2019)](https://old.hotchips.org/hc31/HC31_T6_Rocket.pdf), `To RoCC Accelerator` is located after WB stage.
[Website that illustrates the Rocketchip microarchitecture](https://www.cl.cam.ac.uk/~jrrk2/docs/tagged-memory-v0.1/rocket-core/) would also help.

**Comment by JongyCysec** [comment]
@woojinnn 
Thanks to your sincere answer, I can have a more precise view on how RoCC Accelerator works with host Rocket cpu and how current AI compiler deal with generating code for SoC accelerator.

**Comment by woojinnn** [comment]
@JongyCysec Nice that my answer could help! Since I don't know many things about RoCC Accelerator and compiler, please let me know if you find out interesting stuff!

**Comment by JongyCysec** [comment]
@woojinnn 

Thanks to your answer, I could have checked that ```Rocc``` command is issued at ```writeback``` stage.

Then, I wonder whether it is applied to out-of-order ```Boom``` chip in the same way or not.
At first, I expected ```Boom``` chip can handle in different way and might be able to issue ```Rocc``` command at earlier stage since it manages reorder buffer(```Rob```).


## Boom chip documentation
>According to the documentation of Boom chip(https://docs.boom-core.org/en/latest/sections/execution-stages.html?highlight=rocc),  it is said that 
>- **As BOOM does not store the instruction bits in the ROB, a separate data structure (A “RoCC Shim”) holds the instructions until the RoCC instruction can be committed and the RoCC command sent to the co-processor.**
>- **After issue to RoCC, we track a queue of in-flight RoCC instructions, since we need to translate the logical destination register identifier from the RoCC response into the previously renamed physical destination register identifier.**


In addition, about reorder buffer, 

>### Point of No Return (PNR)
>The point-of-no-return head runs ahead of the ROB commit head, marking the next instruction which might be misspeculated or generate an exception. These include unresolved branches and untranslated memory operations. Thus, the instructions ahead of the commit head and behind the PNR head are guaranteed to be non-speculative, even if they have not yet written back.
>
>Currently the PNR is only used for RoCC instructions. RoCC co-processors typically expect their instructions in-order, and do not tolerate misspeculation. Thus we can only issue a instruction to our co-processor when it has past the PNR head, and thus is no longer speculative.


So, it is not explicitly written in the documentation where ```Rocc``` commands are issued.
But, based on the explanation of the documentation, I expect that they are issued at execute pipeline and managed in separate queue for committing. But it does not tolerate misspeculation like Rocket chip. 

Thanks for the kind words and direction, I appreciate the opportunity to learn more in diverse way.

**Comment by woojinnn** [comment]
1. As you said, I think `Rocc` command is issued at execute stage. [related code](https://github.com/riscv-boom/riscv-boom/blob/1606025fe0e0b10fcc45f4bd609a4fc30458054f/src/main/scala/v4/exu/core.scala#L60)
---
>  I expect that they are issued at execute pipeline and managed in separate queue for committing. But it does not tolerate misspeculation like Rocket chip.

2. I'll search more about it.
[According to the Section 4.1 of Boom Paper](https://carrv.github.io/2020/papers/CARRV2020_paper_15_Zhao.pdf), it says,
```text
Unlike Rocket, SonicBOOM speculatively decodes RoCC instructions and holds their operands in a “RoCC queue” until they pass the architectural commit point, at which time they can be issued into the accelerator
```
I think issuing to the accelerator happens when BOOM core can ensure that misspeculation doesn't occur.

**Comment by JongyCysec** [comment]
Oh, I got it. 

So, the overall order would be 

### Rocket (In-order core)
Issued at ```writeback``` stage. (Architectural commit of operands is guaranteed as it is issued at writeback stage)

### Boom (Out-of-order core)
- Fetch  
- Decode  
  * Put Rocc command in the queue and wait for architectural commit of its operands.
- Execute
  * Issuing the command to Rocc


According to the Boom paper you provided, it is said that **RoCC interface to the accelerator can write data directly into core registers. Additionally, the accelerator can also access ports to the L1, L2, outer memory, and MMU.**

So in my expectation, 
- Register dependency between previous instruction & rocc instruction is resolved by waiting for architectural commit of operands at decode stage.
- Register dependency between afterward instruction & rocc instruction might be resolved by logical register file (=architectural register file) or register renaming.
- And dependency between memory accesses of rocc commands and cpu lw/sw instructions might be resolved by load-store queue. 

To sum up, by resolving the dependencies via additional hardware modules (rob, architectural registers, lsq), Boom chip can apply out-of-order execution design to Rocc commands, too.

**Comment by minseongg** [comment]
> I expect that they are issued at execute pipeline and managed in separate queue for committing.

Yes, the RoCC commands are issued at execute pipeline and manages two internal queues: `rxq` and `rcq`.

- `rxq`: Tracks RoCC commands that have been issued from the ROB, waiting to be issued to the RoCC accelerator.
- `rcq`: Tracks RoCC commands that have been issued to the RoCC accelerator, waiting to be committed.

You can check the implementation in [exu/execution-units/rocc.scala](https://github.com/riscv-boom/riscv-boom/blob/1606025fe0e0b10fcc45f4bd609a4fc30458054f/src/main/scala/v3/exu/execution-units/rocc.scala).

A RoCC command is processed as follows:

1. When a RoCC command is issued from the ROB, it is stored at the head of `rxq` ([code](https://github.com/riscv-boom/riscv-boom/blob/1606025fe0e0b10fcc45f4bd609a4fc30458054f/src/main/scala/v3/exu/execution-units/rocc.scala#L141-L148)).
2. The tail of `rxq` waits for its operands and it is guaranteed to be non-speculative ([code](https://github.com/riscv-boom/riscv-boom/blob/1606025fe0e0b10fcc45f4bd609a4fc30458054f/src/main/scala/v3/exu/execution-units/rocc.scala#L150-L170)). It need to wait until it is guaranteed to be non-speculative because **"RoCC accelerators"** typically require in-order execution and do not support handling mis-speculated commands (e.g., NPU does not support kill signals). Therefore, it is necessary even though BOOM contains reorder buffer and supports out-of-order execution.
3. Once the tail of `rxq` is guaranteed as non-speculative, it is issued to the RoCC accelerator. At the same time, the command is enqueued to the `rcq` ([code](https://github.com/riscv-boom/riscv-boom/blob/1606025fe0e0b10fcc45f4bd609a4fc30458054f/src/main/scala/v3/exu/execution-units/rocc.scala#L177-L191)).
4. When the result of a RoCC command is returned from the RoCC accelerator, it is wrapped with the tail of `rcq` and continues to the next stage ([code](https://github.com/riscv-boom/riscv-boom/blob/1606025fe0e0b10fcc45f4bd609a4fc30458054f/src/main/scala/v3/exu/execution-units/rocc.scala#L237-L259)).

You can check simple figure which illustrates the structure as follows:

<p align="center">
  <img src="https://github.com/user-attachments/assets/2a2fc686-bdea-46a1-8940-9c3ec1c34594" width=50% />
</p>

> But it does not tolerate misspeculation like Rocket chip.

That's correct, but it can issue commands to the RoCC accelerator more quickly because it doesn't need to wait for all previous instructions to be actually committed.

**Comment by JongyCysec** [comment]
@minseongg @woojinnn 
I'm very sorry for frequently asking some specific and bothering questions.

Since rocc touches not only register but also memory region (cache, system memory), I expect that memory access dependency should be resolved somehow.

And I think memory access of rocc command should be considered by load-store queue. 

Then, the range of memory address that rocc accesses matters and how load-store queue can determine such memory range matters. 

Can I ask...
if the range is not explicitly provided in the rocc command, then how can we handle such dependency?

**Comment by minseongg** [comment]
Hmm.. That's very interesting question! Let me explore and answer to the question.

For now, I'm not sure such memory dependency resolve logic is implemented or not.

**Comment by JongyCysec** [comment]
@woojinnn @minseongg 
It's easy to sit down and ask a question. However, it's hard to find the related resources  and reply with an easy-to-understand answer. I always feel sorry and grateful for your sincere efforts.

**Comment by JongyCysec** [comment]
As far as I investigated,

among the extended isa of gemmini, ```mvin```(Move Data From Main Memory to Scratchpad) & ```mvout```(Move Data from Scratchpad to L2/DRAM) instructions matter with regard to memory synchronization with subsequent host cpu memory instructions like ```sw```/```lw```.
(https://github.com/ucb-bar/gemmini?tab=readme-ov-file#mvin-move-data-from-main-memory-to-scratchpad)

So, if host Boom chip cpu encounters such gemmini instructions(```mvin```, ```mvout```), then it needs to synchronize with subsequent memory instructions.

## Solution 1. 
At first, as I described before, 
one possible solution will be notifying the address range that gemmini accesses to ```Load/Store Unit```, expecially ```Load/Store Queue```. 
Then, how to determine/calculate the exact address range from gemmini instruction matters. 


## Solution 2. 
From the decode logic of risc-v Boom chip codebase, it allows to set configuration for extended ISA of Rocc unit.  
https://github.com/riscv-boom/riscv-boom/blob/1606025fe0e0b10fcc45f4bd609a4fc30458054f/src/main/scala/v3/exu/decode.scala#L410-L450

```scala
object RoCCDecode extends DecodeConstants
{
  // Note: We use FU_CSR since CSR instructions cannot co-execute with RoCC instructions
                       //                                                                   frs3_en                        wakeup_delay
                       //     is val inst?                                                  |  imm sel                     |    bypassable (aka, known/fixed latency)
                       //     |  is fp inst?                                                |  |     uses_ldq              |    |  is_br
                       //     |  |  is single-prec                          rs1 regtype     |  |     |  uses_stq           |    |  |
                       //     |  |  |                                       |       rs2 type|  |     |  |  is_amo          |    |  |
                       //     |  |  |  micro-code           func unit       |       |       |  |     |  |  |  is_fence     |    |  |
                       //     |  |  |  |           iq-type  |               |       |       |  |     |  |  |  |  is_fencei |    |  |  is breakpoint or ecall?
                       //     |  |  |  |           |        |       dst     |       |       |  |     |  |  |  |  |  mem    |    |  |  |  is unique? (clear pipeline for it)
                       //     |  |  |  |           |        |       regtype |       |       |  |     |  |  |  |  |  cmd    |    |  |  |  |  flush on commit
                       //     |  |  |  |           |        |       |       |       |       |  |     |  |  |  |  |  |      |    |  |  |  |  |  csr cmd
                       //     |  |  |  |           |        |       |       |       |       |  |     |  |  |  |  |  |      |    |  |  |  |  |  |
  val table: Array[(BitPat, List[BitPat])] = Array(//       |       |       |       |       |  |     |  |  |  |  |  |      |    |  |  |  |  |  |
    CUSTOM0            ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_X  , RT_X  , RT_X  , N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM0_RS1        ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_X  , RT_FIX, RT_X  , N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM0_RS1_RS2    ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_X  , RT_FIX, RT_FIX, N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM0_RD         ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_FIX, RT_X  , RT_X  , N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM0_RD_RS1     ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_FIX, RT_FIX, RT_X  , N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM0_RD_RS1_RS2 ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_FIX, RT_FIX, RT_FIX, N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM1            ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_X  , RT_X  , RT_X  , N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM1_RS1        ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_X  , RT_FIX, RT_X  , N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM1_RS1_RS2    ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_X  , RT_FIX, RT_FIX, N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM1_RD         ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_FIX, RT_X  , RT_X  , N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM1_RD_RS1     ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_FIX, RT_FIX, RT_X  , N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM1_RD_RS1_RS2 ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_FIX, RT_FIX, RT_FIX, N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM2            ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_X  , RT_X  , RT_X  , N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM2_RS1        ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_X  , RT_FIX, RT_X  , N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM2_RS1_RS2    ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_X  , RT_FIX, RT_FIX, N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM2_RD         ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_FIX, RT_X  , RT_X  , N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM2_RD_RS1     ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_FIX, RT_FIX, RT_X  , N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM2_RD_RS1_RS2 ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_FIX, RT_FIX, RT_FIX, N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM3            ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_X  , RT_X  , RT_X  , N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM3_RS1        ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_X  , RT_FIX, RT_X  , N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM3_RS1_RS2    ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_X  , RT_FIX, RT_FIX, N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM3_RD         ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_FIX, RT_X  , RT_X  , N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM3_RD_RS1     ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_FIX, RT_FIX, RT_X  , N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N),
    CUSTOM3_RD_RS1_RS2 ->List(Y, N, X, uopROCC   , IQT_INT, FU_CSR, RT_FIX, RT_FIX, RT_FIX, N, IS_X, N, N, N, N, N, M_X  , 0.U, N, N, N, N, N, CSR.N)
  )
}
```

I'm not sure. But, fields (```use_ldq```, ```use_stq```, ```is_fence```) look quite helpful to resolve our problem. 
In my expectation, memory address range that is applied to ```use_ldq``` or ```use_stq``` is not clear. It may be the size of one word. But it is not appropriate to our situation where we need to protect overall memory address region that matrix resides. 
So, I think ```is_fence``` will be proper to resolve such problem and I expect we can freely modify some fields of decoding logic for our own accelerator ISA specification.

Therefore, if my guess for above decode code is valid, then the developer of ```Rocc``` is responsible to adjust decoding settings properly so that  ```Rocc commands``` become synchronizable with ```Host cpu memory operations```.

**Comment by minseongg** [comment]
I also think it is important to use such `fence` or `flush` instructions properly for memory ordering (Slide 16 of [here](https://carrv.github.io/2020/slides/CARRV2020_slides_3_Xu.pdf) may related). In some test cases in gemmini-rocc-tests ([matmul.c](https://github.com/ucb-bar/gemmini-rocc-tests/blob/dev/bareMetalC/matmul.c), [mvin_mvout.c](https://github.com/ucb-bar/gemmini-rocc-tests/blob/dev/bareMetalC/mvin_mvout.c)), we can find that they are using `gemmini_fence()` and `gemmin_flush()` commands, and actually the `gemmini_fence()` instruction is defined as [a macro](https://github.com/ucb-bar/gemmini-rocc-tests/blob/dev/include/gemmini.h#L297-L298) for `fence` assembly.

And another thing that I have in mind is `dis_hazards` calculation logic in execution unit ([code](https://github.com/riscv-boom/riscv-boom/blob/master/src/main/scala/v3/exu/core.scala#L695-L708)) is also related to the memory ordering..?

I'm not sure for this answer, I think more accurate way is to build and run the simulator with BOOM + gemmini and then analyze how they manages the instructions. Sorry for giving an incomplete answer.

**Comment by JongyCysec** [comment]
Oh, I got it. 
As you have said, ```fence()``` instruction of ```gemmini``` can handle this situation. 
(related issue at gemmini github, https://github.com/ucb-bar/gemmini/issues/318)

Thank you for your answer !!
